subsys/net/lib/aws_iot/src/aws_iot.c

```diff
@@ -0,0 +1,679 @@
+		return -ENOMEM;
+	}
+
+#if defined(CONFIG_AWS_IOT_GET_ACCEPTED_TOPIC_SUBSCRIBE)
+	err = snprintf(get_accepted_topic, sizeof(get_accepted_topic), GET_ACCEPTED_TOPIC,
+		       client_id_buf);
+	if (err != GET_ACCEPTED_TOPIC_LEN) {
+		return -ENOMEM;
+	}
+#endif
+#if defined(CONFIG_AWS_IOT_GET_REJECTED_TOPIC_SUBSCRIBE)
+	err = snprintf(get_rejected_topic, sizeof(get_rejected_topic), GET_REJECTED_TOPIC,
+		       client_id_buf);
+	if (err != GET_REJECTED_TOPIC_LEN) {
+		return -ENOMEM;
+	}
+#endif
+#if defined(CONFIG_AWS_IOT_UPDATE_ACCEPTED_TOPIC_SUBSCRIBE)
+	err = snprintf(update_accepted_topic, sizeof(update_accepted_topic), UPDATE_ACCEPTED_TOPIC,
+		       client_id_buf);
+	if (err != UPDATE_ACCEPTED_TOPIC_LEN) {
+		return -ENOMEM;
+	}
+#endif
+#if defined(CONFIG_AWS_IOT_UPDATE_REJECTED_TOPIC_SUBSCRIBE)
+	err = snprintf(update_rejected_topic, sizeof(update_rejected_topic), UPDATE_REJECTED_TOPIC,
+		       client_id_buf);
+	if (err != UPDATE_REJECTED_TOPIC_LEN) {
+		return -ENOMEM;
+	}
+#endif
+#if defined(CONFIG_AWS_IOT_UPDATE_DELTA_TOPIC_SUBSCRIBE)
+	err = snprintf(update_delta_topic, sizeof(update_delta_topic),
+		       UPDATE_DELTA_TOPIC, client_id_buf);
+	if (err != UPDATE_DELTA_TOPIC_LEN) {
+		return -ENOMEM;
+	}
+#endif
+#if defined(CONFIG_AWS_IOT_DELETE_ACCEPTED_TOPIC_SUBSCRIBE)
+	err = snprintf(delete_accepted_topic, sizeof(delete_accepted_topic),
+		       DELETE_ACCEPTED_TOPIC, client_id_buf);
+	if (err != DELETE_ACCEPTED_TOPIC_LEN) {
+		return -ENOMEM;
+	}
+#endif
+#if defined(CONFIG_AWS_IOT_DELETE_REJECTED_TOPIC_SUBSCRIBE)
+	err = snprintf(delete_rejected_topic, sizeof(delete_rejected_topic),
+		       DELETE_REJECTED_TOPIC, client_id_buf);
+	if (err != DELETE_REJECTED_TOPIC_LEN) {
+		return -ENOMEM;
+	}
+#endif
+	return 0;
+}
+
+#if defined(CONFIG_AWS_FOTA)
+static void aws_fota_cb_handler(enum aws_fota_evt_id evt)
+{
+	struct cloud_backend_config *config = aws_iot_backend->config;
+	struct cloud_event cloud_evt = { 0 };
+
+	switch (evt) {
+	case AWS_FOTA_EVT_DONE:
+		LOG_DBG("AWS_FOTA_EVT_DONE");
+		cloud_evt.type = CLOUD_EVT_FOTA_DONE;
+		cloud_notify_event(aws_iot_backend, &cloud_evt,
+				   config->user_data);
+		break;
+	case AWS_FOTA_EVT_ERROR:
+		LOG_ERR("AWS_FOTA_EVT_ERROR");
+		cloud_evt.type = CLOUD_EVT_ERROR;
+		cloud_notify_event(aws_iot_backend, &cloud_evt,
+				   config->user_data);
+		break;
+	}
+}
+#endif
+
+static int aws_iot_init(const struct cloud_backend *const backend,
+			cloud_evt_handler_t handler)
+{
+	int err;
+
+	backend->config->handler = handler;
+	aws_iot_backend = (struct cloud_backend *)backend;
+
+	err = aws_iot_topics_populate();
+	if (err) {
+		LOG_ERR("aws_topics_populate, error: %d", err);
+		return err;
+	}
+
+#if defined(CONFIG_AWS_FOTA)
+	err = aws_fota_init(&client, "v0.0.0",
+		aws_fota_cb_handler);
+	if (err) {
+		LOG_ERR("aws_fota_init, error: %d", err);
+		return err;
+	}
+#endif
+	return err;
+}
+
+static int aws_iot_ep_topics_set(const struct cloud_backend *const backend,
```

Comment:
How is this function used? I hope it's not called more than once, otherwise, we'll have a memory leak. Possibly, you could `free` the `app_topic_data.list` if it is not NULL, before allocating a new one.

I guess we should also have an `uninit` function that would do the cleanup.

Reply:
This function is called by the application during initialization to pass in a list of subscription topics which will be subscribed to every time the AWS IoT library gets a CONNACK MQTT event. Its intention is to be called only once. However the Cloud API does have an ```uninit`` function which I can attach to a new function in the AWS IoT lib which frees the ```app_topic_data.list```. This will make it possible to change the application specific subscribed topics at run-time. Do you think that the already existing ```uninit``` should be used or shall I make an uninit function spesifically for freeing the ```app_topic_data.list```?

Reply:
I was thinking about re-using the existing `uninit` API - it should clean up any mess that cloud engine made during its operation. I can imagine a situation when we init/unint the cloud engine several times during a single runtime cycle, and in that case, it's important to have a proper cleanup.

And as for calling malloc within `CONNACK MQTT` event handler - are you sure it will happen only once during application runtime? The cloud could disconnect over time for various reasons, I'd imagine to re-connect in such case instead of rebooting entire application. And in such case we'll get another CONNACK event .

Reply:
There are two functions that "manage" the subscription list. ```aws_iot_topics_subscribe``` allocates a list of application spesific topics in the initialization phase of the app(linked through the cloud api) and one internal static function ```topics_subscribe``` which is called every time the aws iot library gets a connack. ```topics_subscribe``` does not allocate memory, it just subscribes to topics depending on the content of aws shadow topics(Kconfigurable) and application spesific topic list (which ```aws_iot_topics_subscribe``` allocates memory for). The memory leak will occur only if the application calls ```cloud_topics_subscribe``` more than once without freeing the previous allocated list.

Reply:
I'm late to the party here, but would be nice to avoid the heap entirely. I know you've mem slabs API a bit, @rlubos . Is that a candidate here?

Reply:
If it's desired to avoid heap usage here, I'd consider static allocation in this particular use-case. It only makes sense to have a single `app_topic_data.list` available at a time, and the only variable here is number of topics registered.

We could introduce a Kconfig option to specify the maximum number of topics allowed, and simply pre-allocate the list in static memory. While a mem slab could also be considered here, I think it has no added value for this use-case - it would only complicate the code a bit, and we would need to apply some limits on it anyway.

Reply:
Agreed, (Kconfigurable sized) static allocation is reasonable.

Reply:
That sounds like a reasonable alternative. Ill make the appropriate changes.